### ISSUE TITLE
No root partition for specified root hash found in '/dev/loop3'.

### ISSUE THREAD

**Issue by protonesso** [systemd/systemd#19200]
Hi! I can't mount OS image with verity data. I use GPT partition table, I expect to mount everything automatically using "Discoverable Partitions Specification". Did I something wrong? What should I do instead.

**systemd version the issue has been seen with**
 > 247 (247.3-3)

**Used distribution**
 > Debian 11

**Linux kernel version used** (`uname -a`)

 > Linux uwu 5.10.0-5-amd64 #1 SMP Debian 5.10.26-1 (2021-03-27) x86_64 GNU/Linux

**CPU architecture issue was seen on**
 > all

**Expected behaviour you didn't see**
 > Mount loop device with verity checking

**Unexpected behaviour you saw**
 > No root partition for specified root hash found in '/dev/loop3'.

**Steps to reproduce the problem**
1. Make a raw disk image with following partition layout:
```
Device        Start      End  Sectors  Size Type
os.img1        2048   264191   262144  128M EFI System
os.img2      264192 31721471 31457280   15G Linux root (x86-64)
os.img3    31721472 42207231 10485760    5G Linux root verity (x86-64)
os.img4    42207232 52428766 10221535  4.9G Linux variable data
```
2. Format verity partition and copy root hash
3. Mount disk image using this command:
```
systemd-dissect /dev/loop3 -m os --root-hash=4ea8e63059b23f006d4ef56486a17e25914b24cdbc461593a4bfb6e7516909a1
```

**Comment by poettering** [comment]
if you specify a root hash of a verity partition we'll use that to search for the root partition and the hash partition: the GPT partition UUID of the root fs should be the first 128bit of the root hash, and the verity partition should use the last 128bit of the root hash as GPT partition UUID.

This is documented here:

https://systemd.io/DISCOVERABLE_PARTITIONS/

This is done so that the root hash pinpoints exactly one combination of partitions in a pool of many and makes things cleary clear.

That said, we certainly could beef things up so that when we operate on a disk image and there are exactly one set of partitions that could possibly be the right ones we are more lenient and ignore the GPT partition UUIDs and just use the ones we find. Hence turning this into an RFE bug

**Comment by protonesso** [comment]
> If this feature is used the partition UUID of the root partition should be the first 128bit of the root hash of the dm-verity hash data, and the partition UUID of this dm-verity partition should be the final 128bit of it, so that the root partition and its verity partition can be discovered easily, simply by specifying the root hash.

How can I do this with frisk or/and any other tools?

**Comment by poettering** [comment]
"frisk"? you mean fdisk? in fdisk: switch to expert mode with "x", and then use "u" to change a partition's uuid

**Comment by protonesso** [comment]
Yeah, fdisk! Stupid keyboard. Thank you for showing me the way to do this. I'll try tomorrow
